### PR TITLE
Add functionality for early stopping rounds.

### PR DESCRIPTION
**Pull request by david-sun-1**
Had a need for early stopping rounds and followed the discussion found here https://github.com/dmlc/XGBoost.jl/issues/158 to implement this version of it.

### CODE REVIEW

File: src/booster.jl [hunk 1]
```diff
@@ -422,10 +421,44 @@ Run `num_round` rounds of gradient boosting on [`Booster`](@ref) `b`.
...
+                 num_round::Integer=1, 
+                 watchlist=Dict("train"=>Xy), 
+                 early_stopping_rounds::Integer=0,
+                 maximize=false,
+                 kw...,
+                 )
+
+    if !isempty(watchlist) && early_stopping_rounds > 0
+        early_stopping_set = collect(Iterators.map(string, keys(watchlist)))[end]
+        println("Will train until there has been no improvement in $early_stopping_rounds rounds.\n")
```
Comment (ExpandingMan) [inline src/booster.jl:434]: This should be an `@info` or `@debug` log, not a `print`.

**Comment by ExpandingMan** [comment]
Thanks for this.

I have a few concerns with this implementation.  The first is that, we had thus far carefully avoided parsing the output messages for obvious reasons.  There is apparently no way around the fact that having this kind of functionality requires it, so I don't think we could disallow it, however, I think that if we are going to do this we should (at the very least) have dedicated log message parsing functions, rather than cramming it into `update!`.  Ideally, we probably should provide the ability to insert arbitrary callbacks between calls to `updateone!`.  That doesn't have to come in this PR, but we should think carefully about what the interface should look like for it to see if we can do it without breaking (I have not thought it through carefully yet myself).  I also don't love the idea of removing the evaluations from the calls to `updateone!` itself, as it could be useful if someone wants to write a custom loop that uses it, though I can certainly see a good argument that it's getting in the way here.  Lastly, shouldn't we have some kind of tolerance for the score comparison?  It would be really nice to be able to do some kind of simple convergence check and stop when that is reached.  In most of my cases I'd be far more likely to use this to do that and I think that is a good use case.

File: src/booster.jl [hunk 2]
```diff
@@ -441,6 +526,13 @@ followed by [`update!`](@ref) for `nrounds`.
 `watchlist` is a dict the keys of which are strings giving the name of the data to watch
 and the values of which are [`DMatrix`](@ref) objects containing the data.
 
+`early_stopping_rounds` if 0, the early stopping function is not triggered. If set to a positive integer, 
+training with a validation set will stop if the performance doesn't improve for k rounds.
+
+`maximize` If early_stopping_rounds is set, then this parameter must be set as well.
```
Comment (ExpandingMan) [inline src/booster.jl:532]: This seems to imply that the user must input the value whenever they input `early_stopping_rounds`.  Maybe better to say something like "only used if `early_stopping_rounds > 0`".

**Comment by ExpandingMan** [comment]
Ok, comments in line.  I think the gist is that, in order not to return a message string from `updateone!` (if anything I think it should return the actual metric, not a string, but this would still be technically breaking), logging will have to be moved into `update!`.  Doesn't seem ideal, but I don't see a lot of options so I think I'm ok with that.

**Comment by wilan-wong-1** [comment]
@ExpandingMan I've updated the early stopping rounds branch to incorporate some fixes. Predominantly, as the watchlist was previously defaulting to a Dict, this is an unordered data structure which meant relying on the ordering of elements in this data structure could cause unintentional behaviour. Modifications have been made to account for this to ensure order is preserved based on user input (if an OrderedDict is provided). 

To ensure backwards compatibility, I've maintained the same xgboost interface but placed a warning if a user inputs a Dict watchlist with more than 1 element and has decided to use the early stopping rounds functionality.

To support early stopping rounds, I've also extended the Booster class to have additional attributes which capture the best round (iteration) and best score as per other XGBoost language implementations. This can then be directly used with the already defined ntree_limit parameter in the predict() method.

I've made additional test cases to also give test coverage to these enhancements.

**Comment by ExpandingMan** [comment]
Thanks, especially for all the testing.  The way I see it the major issue here is the return of the message string, as commented above.  Even if it were not breaking it's a bizarre interface.  There are lots of other ways of handling this.

**Review by ExpandingMan — COMMENTED** [review]
Sorry guys, I just realized after doing the same thing on another PR that nobody could see the comments I wrote.  I hate this entire github feature.

File: src/booster.jl [hunk 3]
```diff
@@ -450,23 +564,51 @@ See [`updateone!`](@ref) for more details.
...
+
+    # We have a watchlist - give a warning if early stopping is provided and watchlist is a Dict type with length > 1
+    if isa(watchlist, Dict)
+        if early_stopping_rounds > 0 && length(watchlist) > 1
+            @warn "Early stopping rounds activated whilst watchlist has more than 1 element. Recommended to provide watchlist as an OrderedDict to ensure deterministic behaviour."
+        end
+    end
+
+    if isempty(watchlist) && early_stopping_rounds > 0
+        @warn "Early stopping is ignored as provided watchlist is empty."
```
[the remark this answers is not in the document]
Comment (wilan-wong-1) [reply src/booster.jl]: @ExpandingMan I have addressed the comments in the latest commit. Can you take a look?

**Comment by ExpandingMan** [comment]
This looks good to me, thanks!  Let me give it a more thorough look when I get a chance, and hopefully I can merge.

**Comment by ExpandingMan** [comment]
I haven't forgotten this but I haven't gone through it yet... feel free to ping me if I don't get back to it this week.

**Comment by ExpandingMan** [comment]
Looks good to me, thanks for this.

**Pull request merged**